reward/staking_manager.go

```diff
@@ -39,66 +39,101 @@ type blockChain interface {
 
-	if cachedStakingInfo := sm.sic.get(stakingBlockNumber); cachedStakingInfo != nil {
-		logger.Debug("StakingInfoCache hit.", "Block number", blockNum, "number of staking block", stakingBlockNumber)
+	// Get staking info from cache
+	if cachedStakingInfo := sm.stakingInfoCache.get(stakingBlockNumber); cachedStakingInfo != nil {
+		logger.Debug("StakingInfoCache hit.", "blockNum", blockNum, "staking block number", stakingBlockNumber, "stakingInfo", cachedStakingInfo)
 		return cachedStakingInfo
 	}
 
-	stakingInfo, err := sm.updateStakingCache(stakingBlockNumber)
+	// Get staking info from DB
+	if sm.stakingInfoDB != nil {
+		if storedStakingInfo, err := sm.stakingInfoDB.ReadStakingInfo(stakingBlockNumber); storedStakingInfo != nil && err == nil {
+			s, ok := storedStakingInfo.(*StakingInfo)
+			if ok {
+				logger.Debug("StakingInfoDB hit.", "blockNum", blockNum, "staking block number", stakingBlockNumber, "stakingInfo", s)
+				return s
+			} else {
+				logger.Warn("Cannot convert to StakingInfo. Wrong data is written in DB.")
+			}
+		} else {
+			logger.Warn("Failed to get staking info from db.", "blockNum", blockNum, "err", err)
+		}
+	}
+
+	// Get staking info from block header and updates it to cache and db
+	calcStakingInfo, err := sm.updateStakingInfo(stakingBlockNumber)
 	if err != nil {
-		logger.Error("Failed to get stakingInfo", "Block number", blockNum, "number of staking block", stakingBlockNumber, "err", err)
+		logger.Error("Failed to get stakingInfo", "blockNum", blockNum, "staking block number", stakingBlockNumber, "err", err, "stakingInfo", calcStakingInfo)
 		return nil
 	}
 
-	logger.Debug("Complete StakingInfoCache update.", "Block number", blockNum, "number of staking block", stakingBlockNumber)
-	return stakingInfo
+	logger.Debug("Get stakingInfo from header.", "blockNum", blockNum, "staking block number", stakingBlockNumber, "stakingInfo", calcStakingInfo)
+	return calcStakingInfo
 }
 
 func (sm *StakingManager) IsActivated() bool {
 	return sm.isActivated
 }
 
-// updateStakingCache updates staking cache with a stakingInfo of a given block number.
-func (sm *StakingManager) updateStakingCache(blockNum uint64) (*StakingInfo, error) {
-	stakingInfo, err := sm.ac.getStakingInfoFromAddressBook(blockNum)
+// updateStakingInfo updates staking info in cache and db created from given block number.
+func (sm *StakingManager) updateStakingInfo(blockNum uint64) (*StakingInfo, error) {
+	stakingInfo, err := sm.addressBookConnector.getStakingInfoFromAddressBook(blockNum)
 	if err != nil {
 		return nil, err
 	}
 	sm.isActivated = true
-	sm.sic.add(stakingInfo)
 
-	logger.Info("Add a new stakingInfo to the stakingInfoCache", "stakingInfo", stakingInfo)
+	// update cache
+	sm.stakingInfoCache.add(stakingInfo)
+
+	// update db
+	if sm.stakingInfoDB != nil {
+		err := sm.stakingInfoDB.WriteStakingInfo(blockNum, stakingInfo)
+		if err != nil {
+			logger.Warn("Failed to write staking info to db.", "err", err)
+		}
+	}
```

Comment:
Why doesn't it return an error when it failed to write data to DB?

Reply:
I changed it to return the error, after your comment.
However, the returned error is not used right now. The main purpose of calling `updateStakingInfo` is to get staking info from header. The write failure is not a critical issue in this situation.

Reply:
"staking info" was stored only in cache, but now it is stored both of DB and cache.
It doesn't make sense for me that data is stored in cache, but not in DB.